lib/capybara/selector/definition/button.rb

```diff
@@ -21,8 +21,8 @@
 
       btn_xpath = btn_xpath[locator_matchers |
                             XPath.string.n.is(locator) |
-                            XPath.descendant(:img)[XPath.attr(:alt).is(locator)]
-                           ] + locate_label(locator).descendant(btn_xpath)
+                            XPath.descendant(:img)[XPath.attr(:alt).is(locator)]]
+      btn_xpath += locate_label(locator).descendant(btn_xpath)
```

Comment:
@twalpole your hunch was correct, this seems to resolve the issue.

Reply:
Hmmm -- that doesn't seem right to me since it would mean the locator would have to match in the label and also match against the actual button element. I'll have to look at the XPath that's actually being generated -- probably won't have time to dig into that until next weekend

Reply:
I've pushed up test coverage for this use case.

Reply:
Reading the spec -- since &lt;label elements can't be associated with aria-role button elements it probably doesn't make sense to support them like that

Reply:
@twalpole oddly enough, the feature work that uncovered this bug is related to `label [role="button"]` elements, inspired by the structure mentioned in [_A styled accessible file upload](https://accessabilly.com/a-styled-accessible-file-upload/):

```html
<label for="fileupload" id="buttonlabel">
  <span role="button" aria-controls="filename" tabindex="0">upload file</span>
</label>
```

Which doesn't get you "button" behavior out of the box, but the structure is accessible and flexible enough for more control.

Would you be able to link to the part of the specification you mentioned?

Reply:
In that case your label element should be associated with an id of "fileupload" - spec is https://www.w3.org/TR/html52/sec-forms.html#the-label-element - relevant portion is "The for attribute may be specified to indicate a form control with which the caption is to be associated. If the attribute is specified, the attribute’s value must be the ID of a labelable element" where https://www.w3.org/TR/html52/sec-forms.html#labelable-element defines labelable elements

Reply:
This is great information to have, thanks for the links.

With that in mind, I don't think the implementation has been influenced at all by this use case, but this PR _does_ introduce test coverage for it.

Are you suggesting that we omit the tests that cover that behavior, namely:

* https://github.com/teamcapybara/capybara/pull/2427/files#diff-3497e3e50e487b0afcc2d1da0a08faf11991dc5986120253042adea31f3cfb6fR465-R468
* https://github.com/teamcapybara/capybara/pull/2427/files#diff-59d5da4957aa6a81605ec0fc1e0cda49c9734a63e7f099efcdc2efa69713b1bdR50
* https://github.com/teamcapybara/capybara/pull/2427/files#diff-59d5da4957aa6a81605ec0fc1e0cda49c9734a63e7f099efcdc2efa69713b1bdR58
* https://github.com/teamcapybara/capybara/pull/2427/files#diff-59d5da4957aa6a81605ec0fc1e0cda49c9734a63e7f099efcdc2efa69713b1bdR101
* https://github.com/teamcapybara/capybara/pull/2427/files#diff-59d5da4957aa6a81605ec0fc1e0cda49c9734a63e7f099efcdc2efa69713b1bdR109

Reply:
The question really is whether `it 'should be true for a role=button within a label when enable_aria_role: true' do
    expect(@session).to have_button('role=button within label', enable_aria_role: true)
  end` should actually find the element and pass. I'm leaning towards that it shouldn't since the `label` element isn't actually associated with the 'role=button' element (clicking the label element wouldn't actually do anything with regards to the `role=button` element)

Reply:
> since the `label` element isn't actually associated with the 'role=button' element (clicking the label element wouldn't actually do anything with regards to the `role=button` element)

```html
<label for="fileupload" id="buttonlabel">
  <span role="button" aria-controls="filename" tabindex="0">upload file</span>
</label>
<input type="file" id="fileupload">
```

In this [example HTML's case](https://accessabilly.com/a-styled-accessible-file-upload/),  would `click_on "upload file"` resolve to the `span[role="button"]` element, and click within the `label` element to trigger the `input[type="file"]`?

Reply:
In this PR's fixture HTML, `click_on "role=button within label element"` would do nothing, but `click_on "with other text"` would click the `[role=button]` and click the `label`:

```html
<label>
  role=button within label element
  <span role="button">with other text</span>
</label>
```

I can see that this is a use case that is odd and might be worth avoiding.

Reply:
In the exmaple HTML case it would work because 'upload file' would be found as the text in the "button" and the span would be clicked (which is where the click handler should be attached if the role is set as button.  I don't believe the `click_on "role=button within label element`  should work because the label element that contains the text isn't actually associated with the span (other than as a parent/child) - so if the user clicks inside the label but outside the span nothing will actually happen.

Reply:
I've created a JSFiddle: https://jsfiddle.net/u7x1s5qr/ with the following HTML:

```html
<label for="file-upload">
  role=button within label element
  <span role="button" tabindex="0">
    with other text
  </span>
</label>
<input id="file-upload" type="file" hidden>
```

> so if the user clicks inside the label but outside the span nothing will actually happen.

In the JSFiddle, you can click on both the "role=button within label element" and the "with other text" to open the file dialog.

I would expect a Capybara call to `click_on "with other text"` (or `click_button`) to work.

> I don't believe the click_on "role=button within label element should work/find a button because the label element that contains the text isn't actually associated with the span (other than as a parent/child) 

As far as Capybara's behavior goes, I'd expect that a call to `click_on "role=button within label element"` would fail. This is somewhat different from what would happen by actually clicking on the `role=button within label element` text, but understanding that the implementation of `click_on` is an alias for `click_button` and `click_link`, it makes sense.

Reply:
The difference is that clicking on the `label` element works because it's associated with the actual `<input type="file">` element, not because it has the aria role button in it.  `find(:label, 'role=button within label element').click` will work too if you remove the `<span role="button"` element

Reply:
I see, thanks for clarifying.

With that in mind, what needs to be done in the short term to push this over the finish line?

In addition to this test, there was also [a surprising CI failure](https://travis-ci.org/github/teamcapybara/capybara/jobs/747208490#L598).